concurrent/concurrency_test.go

```diff
@@ -0,0 +1,239 @@
+	}
+}
+
+// before calling bench, remember to set disjunction
+func benchMembero(b *testing.B, listsize int) {
+	membero := memberOfLargeList(listsize)
+	var r []*ast.SExpr
+	for n := 0; n < b.N; n++ {
+		r = micro.Run(-1, membero)
+	}
+	result = r
+}
+
+func BenchmarkMembero10000(b *testing.B) {
+	disjunction = disjPlus
+	benchMembero(b, 10000)
+}
+
+func BenchmarkConcMembero10000(b *testing.B) {
+	disjunction = DisjPlus
+	benchMembero(b, 10000)
+}
+
+func BenchmarkMembero50000(b *testing.B) {
+	disjunction = disjPlus
+	benchMembero(b, 50000)
+}
+
+func BenchmarkConcMembero50000(b *testing.B) {
+	disjunction = DisjPlus
+	benchMembero(b, 50000)
+}
+
+func mapODoubleUnrolled(f func(*ast.SExpr, *ast.SExpr) micro.Goal, list *ast.SExpr) func(*ast.SExpr) micro.Goal {
+	goals := []func(*ast.SExpr) micro.Goal{}
+	for {
+		if list == nil {
+			break
+		}
+		car, cdr := list.Car(), list.Cdr()
+		goals = append(goals, func(x *ast.SExpr) micro.Goal {
+			return f(x, car)
+		})
+		list = cdr
+	}
+	n := len(goals)
+	return func(x *ast.SExpr) micro.Goal {
+		gs := make([]micro.Goal, n+1)
+		vars := make([]*ast.SExpr, n)
+		var cons *ast.SExpr = nil
+		for i := 0; i < n; i++ {
+			v := ast.NewVariable("")
+			vars[i] = v
+			gs[n-i] = goals[n-i-1](v)
+			cons = ast.Cons(v, cons)
+		}
+		gs[0] = micro.EqualO(x, cons)
+		return conjunction(gs...)
+	}
+}
+
+func equalLargeList(n int) func(*ast.SExpr) micro.Goal {
+	mapo := mapODoubleUnrolled(micro.EqualO, largeList(n))
+	return func(q *ast.SExpr) micro.Goal {
+		return mapo(q)
+	}
+}
+
+// before calling bench, remember to set conjunction
+func benchMapo(b *testing.B, listsize int) {
+	membero := equalLargeList(listsize)
+	var r []*ast.SExpr
+	for n := 0; n < b.N; n++ {
+		r = micro.Run(-1, membero)
+	}
+	result = r
+}
+
+func conjPlus(gs ...micro.Goal) micro.Goal {
+	if len(gs) == 0 {
+		return micro.SuccessO
+	}
+	if len(gs) == 1 {
+		return gs[0]
+	}
+	g1 := gs[0]
+	g2 := conjPlus(gs[1:]...)
+	return func() micro.GoalFn {
+		return func(s *micro.State) micro.StreamOfStates {
+			g1s := g1()(s)
+			return micro.Bind(g1s, g2)
+		}
+	}
+}
+
+func BenchmarkMapo10000(b *testing.B) {
+	conjunction = conjPlus
+	benchMapo(b, 10000)
+}
+
+func BenchmarkConcMapo10000(b *testing.B) {
+	conjunction = ConjPlus
+	benchMapo(b, 10000)
+}
+
+func largeListHeadFail(n int) *ast.SExpr {
+	ints := make([]*ast.SExpr, n)
+	for i := 0; i < n-1; i++ {
+		ints[i] = ast.NewInt(int64(i))
+	}
+	ints[0] = ast.NewInt(1)
+	return ast.NewList(ints...)
+}
+
+func equalFailHeadLargeList(n int) func(*ast.SExpr) micro.Goal {
+	list := largeList(n)
+	mapo := mapODoubleUnrolled(micro.EqualO, list)
+	list2 := largeListHeadFail(n)
+	return func(*ast.SExpr) micro.Goal { return mapo(list2) }
+}
+
+// before calling bench, remember to set conjunction
+func benchMapoFailHead(b *testing.B, listsize int) {
+	membero := equalFailHeadLargeList(listsize)
+	var r []*ast.SExpr
+	for n := 0; n < b.N; n++ {
```

Comment:
I think right before the `for` you have to do `b.ResetTimer()` if I remember correctly